okhttp/src/nonJvmMain/kotlin/okhttp3/Headers.kt

```diff
@@ -0,0 +1,118 @@
+/*
+ *  Licensed to the Apache Software Foundation (ASF) under one or more
+ *  contributor license agreements.  See the NOTICE file distributed with
+ *  this work for additional information regarding copyright ownership.
+ *  The ASF licenses this file to You under the Apache License, Version 2.0
+ *  (the "License"); you may not use this file except in compliance with
+ *  the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package okhttp3
+
+import okhttp3.internal.commonAdd
+import okhttp3.internal.commonAddAll
+import okhttp3.internal.commonAddLenient
+import okhttp3.internal.commonBuild
+import okhttp3.internal.commonEquals
+import okhttp3.internal.commonGet
+import okhttp3.internal.commonHashCode
+import okhttp3.internal.commonHeadersGet
+import okhttp3.internal.commonHeadersOf
+import okhttp3.internal.commonIterator
+import okhttp3.internal.commonName
+import okhttp3.internal.commonNewBuilder
+import okhttp3.internal.commonRemoveAll
+import okhttp3.internal.commonSet
+import okhttp3.internal.commonToHeaders
+import okhttp3.internal.commonToString
+import okhttp3.internal.commonValue
+import okhttp3.internal.commonValues
+
+@Suppress("NAME_SHADOWING")
+actual class Headers internal actual constructor(
+  internal actual val namesAndValues: Array<String>
+) : Iterable<Pair<String, String>> {
+  actual operator fun get(name: String): String? = commonHeadersGet(namesAndValues, name)
+
+  actual val size: Int
+    get() = namesAndValues.size / 2
+
+  actual fun name(index: Int): String = commonName(index)
+
+  actual fun value(index: Int): String = commonValue(index)
+
+  actual fun names(): Set<String> {
+    // TODO check ok with unsorted set
```

Comment:
Yuck yuck yuck. Particularly ugly ’cause the order is shuffled from hashing, right?

Can we put an orderedSet in `internal`? I hate nondeterministic things.

Reply:
Ahhh, no so the issue is that this Set is case-sensitive and that’s incorrect. `Cookie` and `cookie` both show up in the result, and that’s strictly incorrect.

Reply:
Failing test or tracking bug for this?

Reply:
```
    return (0 until size step 2).map { name(it) }.distinctBy { it.lowercase() }.toSet()
```